### PR TITLE
[JAX] Optimize MoE block

### PR DESCRIPTION
**Pull request by jberchtold-nvidia**
# Description

Improves performance of the MoE block by exposing support for quantization, removal of unnecessary masking overheads, and support for less memory usage via a reduced receive capacity in TE EP

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

- Direct support for MXFP8 quantization in the MoE block along with corresponding tests
- Removal of additional overheads like `jnp.where` masking that wasn't required as TE EP and grouped GEMM are group-aware
- Support for a reduced receive capacity and integration with TE EP's overflow detection

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tests/jax/test_te_ep_moe.py [hunk 1]
```diff
@@ -108,18 +109,21 @@ def _read_mp_options():
 
 from transformer_engine_jax import get_device_compute_capability
 
-# Grouped GEMM in the MoE custom_vjp requires Blackwell (sm_100+). The
-# TE EP NCCL primitives themselves need SM>=90, but the FFN body uses
-# grouped_gemm, so the file as a whole gates on sm_100+.
-if get_device_compute_capability(0) < 100:
+# TE EP NCCL primitives need SM>=90
+if get_device_compute_capability(0) < 90:
```
Comment (tdophung) [inline tests/jax/test_te_ep_moe.py]: did this changed? (the grouepd_gemm allows sm 90 now?)

File: transformer_engine/jax/moe.py [hunk 2]
```diff
@@ -160,7 +214,7 @@ def _te_ep_assert_compatible_bootstrap(
         or hidden_dim != b_hidden
         or ep_size != b_ep_size
         or max_tokens_per_rank > b_max_tpr
-        or recv_capacity_per_rank > b_recv_pr
+        or recv_capacity_per_rank != b_recv_pr
```
Comment (tdophung) [inline transformer_engine/jax/moe.py]: I don't recall why we change this check? Was it always this way that we needed `recv_capacity_per_rank` to be the same as `b_recv_pr`?

File: transformer_engine/jax/flax/moe.py [hunk 3]
```diff
@@ -176,7 +185,6 @@ def __call__(self, inputs: Array) -> Tuple[Array, Optional[Array], Array]:
         ----------
         inputs : jnp.ndarray
             ``[batch, sequence, hidden]``.
-
```
Comment (tdophung) [inline transformer_engine/jax/flax/moe.py:179]: nit: might be accidentally breaking the numpydoc generation

File: transformer_engine/jax/flax/moe.py [hunk 4]
```diff
@@ -253,12 +259,39 @@ def __call__(self, inputs: Array) -> Tuple[Array, Optional[Array], Array]:
             )
 
         ep_axis = get_active_resource_axis("ep_resource")
+        mesh = _get_mesh()
```
Comment (tdophung) [inline transformer_engine/jax/flax/moe.py]: `if mesh is None or mesh.empty`

File: transformer_engine/jax/moe.py [hunk 5]
```diff
@@ -369,25 +462,21 @@ def _ffn_bwd_per_shard(
...
+    d_wo_bias = tex.grouped_dbias(d_eo_2d, group_sizes) if has_bias else None
 
     act_fn = _convert_to_activation_function(activation_type)
     if apply_topk_weights_early:
-        # intermediate' = intermediate * w * mask. Split the cotangent
-        # across both factors before the activation bwd consumes it. Padded
-        # recv slots may still be NaN in the saved activation residuals, so
-        # use zero-filled residuals on inactive rows before the activation VJP.
+        # intermediate' = intermediate * w. The subsequent dgrad and EP
+        # operations consume group_sizes, so they skip padded ragged rows.
```
Comment (tdophung) [inline transformer_engine/jax/moe.py]: I thought the reason is that although group_sizes already include the intra-expert paddings (to round up to align_size), and this was confirmed in nccl ep kernels that all the padding are 0s, hencee we can delete the jnp.where here? not that dgrad and EP consum group_sizes so they skip padded ragged rows.  

We can remove the `jnp.where`s but we should fix the comments to be more accurate.

File: transformer_engine/jax/moe.py [hunk 6]
```diff
@@ -852,39 +924,35 @@ def _moe_bwd_rule(
...
         # combine_fwd consumed already-weighted expert_outputs; the recv_w
         # cotangent flows through the early-weighting step inside the FFN bwd.
         d_expert_outputs = grad_pre_combine
         d_recv_w_from_combine = jnp.zeros_like(ctx.recv_topk_weights)
     else:
-        # Reverse the late-weighting multiply. Padded expert-major rows are
-        # part of the physical grouped-GEMM ranges, so write literal zero
-        # cotangents for inactive rows instead of relying on NaN * 0.
+        # Reverse the late-weighting multiply. The subsequent dgrad and EP
+        # operations consume group_sizes, so they skip padded ragged rows.
```
Comment (tdophung) [inline transformer_engine/jax/moe.py]: Fix this comment also (or just remove ccompletely). We do not need to write comments about reversing a change because I don't think it makes sense for people reading this comment without knowing about the previous code

File: transformer_engine/jax/moe.py [hunk 7]
```diff
@@ -282,45 +388,35 @@ def _ffn_fwd_per_shard(
...
         # the downstream wo GEMM is linear so this is equivalent to the
-        # late-weighting path. Padded recv slots can contain uninitialized
-        # data, so overwrite inactive rows with literal zeros instead of
-        # relying on multiplication by a zero mask (IEEE NaN * 0 = NaN).
-        # ``w_b`` is cast to ``intermediate.dtype`` so the multiply doesn't
-        # promote expert_outputs above the EP buffer's element width.
-        w_b = recv_w_flat[:, None].astype(intermediate.dtype)
-        active = (recv_w_flat != 0)[:, None]
-        intermediate = jnp.where(active, intermediate * w_b, jnp.zeros_like(intermediate))
+        # late-weighting path. Grouped GEMM skips padding automatically.
```
Comment (tdophung) [inline transformer_engine/jax/moe.py]: Same with this comment. Grouped GEMM skips the overallocated tail, but does not skip the `align__size` paddings, which are 0 so it does not affect the output

File: transformer_engine/jax/moe.py [hunk 8]
```diff
@@ -65,6 +63,62 @@
 _ALIGN_SIZE = 128
 
 
+def get_moe_recv_capacity_per_rank(
```
Comment (tdophung) [inline transformer_engine/jax/moe.py:66]: This essentially do the same thing as `compute_worst_case_recv_pr`, so we can replace that with this.

Comment (tdophung) [reply transformer_engine/jax/moe.py:66]: in test_te_ep_moe.py. We should replace this anywayys so that the test cases covers this function

File: transformer_engine/jax/flax/moe.py
Comment (tdophung) [inline transformer_engine/jax/flax/moe.py]: wi`'s trailing dim is now `2 * intermediate_size` but is still labelled `"mlp"`, while
`wi_0_bias`/`wi_1_bias` use `"mlp"` for a dim of size `intermediate_size`. If a user's rules map `mlp` to a sharded axis, the fused kernel is split across the gate/up boundary. Ex: if this is sharded by 2, then would one shard gets all of gate and the other all of up proj?

Comment (tdophung) [reply transformer_engine/jax/flax/moe.py]: I see that in `layernorm_mlp` uses `act` as a separate axis but not sure if this is the right move.

File: tests/jax/test_te_ep_moe.py [hunk 9]
```diff
@@ -596,7 +623,7 @@ def test_forward(self, mesh, config):
             np.asarray(jax.device_get(out_ref)).astype(np.float32),
             atol=FWD_ATOL,
             rtol=FWD_RTOL,
```
Comment (tdophung) [inline tests/jax/test_te_ep_moe.py]: should we have different FWD_ATOL and FWD_RTOL for mxfp8 and fp16?

File: tests/jax/test_te_ep_moe.py (hunk 1 above)
Comment (jberchtold-nvidia) [reply tests/jax/test_te_ep_moe.py]: Not yet, this will fallback to the older grouped GEMM V1 impl, but that does not break the TE MoEBlock. Once sm90 is supported for v2 in TE/JAX, it will use v2 instead. But in the meantime, still probably good to enable sm90 tests so we get EP + rest of MoEBlock testing support on Hopper.

I've updated the comment to just say TE EP NCCL primitives need SM>=90

File: transformer_engine/jax/flax/moe.py (hunk 3 above)
Comment (jberchtold-nvidia) [reply transformer_engine/jax/flax/moe.py:179]: Fixed

File: transformer_engine/jax/moe.py (hunk 5 above)
Comment (jberchtold-nvidia) [reply transformer_engine/jax/moe.py]: I've updated the comment to clarify the two types of padding and how they are handled

File: transformer_engine/jax/moe.py (hunk 6 above)
Comment (jberchtold-nvidia) [reply transformer_engine/jax/moe.py]: Comment removed

File: transformer_engine/jax/moe.py (hunk 7 above)
Comment (jberchtold-nvidia) [reply transformer_engine/jax/moe.py]: Updated comment

File: transformer_engine/jax/moe.py (hunk 2 above)
Comment (jberchtold-nvidia) [reply transformer_engine/jax/moe.py]: Reverted back to `>`, this change was accidental. Good catch

File: tests/jax/test_te_ep_moe.py (hunk 9 above)
Comment (jberchtold-nvidia) [reply tests/jax/test_te_ep_moe.py]: Good point, these had pretty relaxed tolerances. I've made the tolerances stricter and split BF16 and MXFP8

File: transformer_engine/jax/flax/moe.py (hunk 3 above)
Comment (jberchtold-nvidia) [reply transformer_engine/jax/flax/moe.py:179]: Fixed, thanks

File: transformer_engine/jax/moe.py (hunk 8 above)
Comment (jberchtold-nvidia) [reply transformer_engine/jax/moe.py:66]: Good point, I've updated the tests to use `get_moe_recv_capacity_per_rank` and deleted the test method `compute_worst_case_recv_pr`. 

The current test setup does EP bootstrap once with the worst-case capacity using a pytest fixture that caches the bootstrap for all tests.

In this PR, I've kept that same setup. But I've added a TODO for a follow-up to refactor these tests. I think refactoring it properly without introducing overhead isn't a super trivial change, as presumably this ep_boostrap logic was cached across tests to avoid the overhead of each test doing NCCL EP bootstrap+teardown itself.
But I will do it as part of a follow-up

**Comment by tdophung** [comment]
LGTM pending CI

**Review by tdophung — DISMISSED** [review]

**Review by tdophung — APPROVED** [review]
